functions/src/index.ts

```diff
@@ -166,3 +177,141 @@ export const loadFirestore = functions.https.onRequest(
     response.status(200).send();
   }
 );
+
+/*
+ * Listen for changes to the User document
+ * and initiate the appropriate Glean ping(s)
+ */
+export const handleUserChangesImpl = async function (
+  change: Change<DocumentSnapshot>,
+  context: EventContext
+): Promise<boolean> {
+  const userID = context.params.userID;
+  const rallyID = await getRallyIdForUser(userID);
+  if (!rallyID) {
+    // Without Rally ID, we can't make any Glean pings
+    // This is bad and should be flagged for inspection
+    throw new Error(
+      `Unable to obtain Rally ID for user ID ${userID}. Aborting Glean ping process.`
+    );
```

Comment:
This exception is currently throwing when I try to delete a user (it looks like the website needs to call this from "Leave Rally" on the website):
```
>  {"severity":"WARNING","message":"Snapshot has no readTime. Using now()"}
⚠  functions: Error: Unable to obtain Rally ID for user ID jg7G6Q5vtdkOzFq2KLjqhnZgLXa2. Aborting Glean ping process.
    at handleUserChangesImpl (file:///Users/mozilla/src/rally-web-platform/functions/lib/index.js:149:15)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async runFunction (/Users/mozilla/src/rally-web-platform/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:533:9)
    at async runBackground (/Users/mozilla/src/rally-web-platform/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:545:5)
    at async processBackground (/Users/mozilla/src/rally-web-platform/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:528:5)
    at async invokeTrigger (/Users/mozilla/src/rally-web-platform/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:610:13)
    at async handleMessage (/Users/mozilla/src/rally-web-platform/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:702:9)
⚠  Your function was killed because it raised an unhandled error.
```

I used the Firebase emulator to delete the user for now, maybe the trigger that deletes the user's docs in Firestore is running first?

![image](https://user-images.githubusercontent.com/61412/164564421-28ba34b9-d054-4878-a91e-507b09c2fad4.png)

Reply:
Oh I think I see, `getRallyIdForUser(userId)` is called here (which is normally fine) and it's going and fetching the document, but if it's already been deleted then you want to use the document in the `change` object being passed: `change.before.data()`.